compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -515,6 +516,73 @@ SortOp::reifyResultShapes(OpBuilder &b,
       .reifyResultShapes(b, reifiedReturnShapes);
 }
 
+namespace {
+struct RemoveUnusedSortOpResults
+    : public OpRewritePattern<IREE::LinalgExt::SortOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(IREE::LinalgExt::SortOp sortOp,
+                                PatternRewriter &rewriter) const override {
+    // To avoid problems in dispatches associated with unused results, prune
+    // them here.
+    Location loc = sortOp->getLoc();
+    auto operands = sortOp.getOutputs();
+    auto results = sortOp.getResults();
+    unsigned numRes = sortOp.getNumResults();
+
+    // If there are no results, bufferization has happened.
+    // Do not try to prune results anymore.
+    if (numRes == 0) {
+      return failure();
+    }
+
+    // If result or associated block arg is used, do not erase.
+    Block &block = sortOp.getRegion().front();
+    auto blockArgs = block.getArguments();
+    SmallVector<Value> usedBlockArgs, usedOperands, usedResults;
+    SmallVector<Type> usedResultTypes;
+    llvm::BitVector eraseArg(numRes * 2, false);
+    for (unsigned idx : llvm::seq<unsigned>(numRes)) {
+      if (!results[idx].use_empty() || !blockArgs[2 * idx].use_empty() ||
+          !blockArgs[2 * idx + 1].use_empty()) {
+        usedOperands.push_back(operands[idx]);
+        usedResults.push_back(results[idx]);
+        usedResultTypes.push_back(results[idx].getType());
+        continue;
+      }
+      eraseArg.set(2 * idx, 2 * idx + 2);
+    }
+
+    // Bail out if no pruning required.
+    if (numRes == usedOperands.size()) {
+      return failure();
+    }
```

Comment:
optional nit: `eraseArg.none()` matches the comment better. `numRes == usedOperands.size()` adds a level of logic IMO.